src/ir/ExpressionAnalyzer.cpp

```diff
@@ -443,6 +338,30 @@ bool ExpressionAnalyzer::flexibleEqual(Expression* left,
       }
       return true;
     }
+
+    bool compareNodes(Expression* left, Expression* right) {
+      if (left->_id != right->_id) {
+        return false;
+      }
+      switch (left->_id) {
+#include "compare-expressions.generated.h"
+        default: {
+          WASM_UNREACHABLE("unknown expression");
+        }
+      }
+      return true;
+    }
+
+    bool compareNames(Name left, Name right) {
+      auto iter = rightNames.find(left);
+      // If it's not found, that means it was defined out of the expression
+      // being compared, in which case we can just treat it literally - it
+      // must be exactly identical.
+      if (iter != rightNames.end()) {
+        left = iter->second;
+      }
```

Comment:
We want `(block $x)` to compare the same as `(block $y)`, that is, to see things as equal if they are isomorphic, ignoring the concrete name choices. So we keep a map of names on the left to names on the right, and then when we see `(br $x)` vs `(br $y)`, we check if `$x` means "the same" as `$y` (which in those blocks it would).